### ISSUE TITLE
Some sites not getting embedly metadata injected via middleware

### ISSUE THREAD

**Issue by pdehaan** [mozilla/activity-streams#284]
I've seen this oddity in the past, but never clued in to the "why", but thanks to @k88hudson's #283 debugger, this is starting to make more sense.

I have 3 items in my browsing history. One bugzilla bug, one Yahoo! search, and one npm package. Two out of the three items don't have the embedly responses (with npm being the only success story).

![new_tab_and_ _ _static_and_inbox__2_223__-_pdehaan_mozilla_com_-_mozilla_mail](https://cloud.githubusercontent.com/assets/557895/13652319/577a5e34-e600-11e5-859e-5af8fb97673b.png)

As close as I can figure, this is because both bugzilla and Yahoo! use query params. And I'm almost certain this is my fault from some problem with the URL sanitizer. Specifically, somewhere around [/content-src/lib/embedly-middleware.js:47](https://github.com/mozilla/activity-streams/blob/0c1ce94335080a5dbcc8e7fbce7bde01cd81bbd4/content-src/lib/embedly-middleware.js#L47). I suspect we're trying to look for the sanitized URLs in the original dataset which wouldn't work, which is why the two URLs wouldn't get any of the embedly data merged back in. But there is probably something else happening incorrectly, as it seems like we should be passing the `?id=` and `?p=` keys through to Embedly, but don't seem to be (see [/content-src/lib/utils.js:41-43](https://github.com/mozilla/activity-streams/blob/0c1ce94335080a5dbcc8e7fbce7bde01cd81bbd4/content-src/lib/utils.js#L41-L43)).

But this sort of explains why some sites like Bugzilla fall back to using the `title` instead of the `provider_name` or `provider_title` (which should have been injected by Embedly middleware, but isn't).

```js
{
  "TopSites": {
    "rows": [
      {
        "url": "https://bugzilla.mozilla.org/show_bug.cgi?id=1255136",
        "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAC2klEQVQ4jYWSbyzUcRzHvxeai8r050GPcMIdd3H3c/fzu8M5fzZdx4XrGGKWIazj/C0zI6Qdskg1Tofz967Lv9QWSmvZanrSVls9aOtJra1FKSa9e1Qo6f3svX0+r73f25uQbRQ6ySB0RBpGCCHSYTpOZqGWt7snitsizkYfMkpDZgkcy6lXetD9wkzGLPy8LUA+yUA+Kin65ZnhwBXpgPjjqcqIvpg6ZpnpEEFbEHTxnwDZiGRBahOvEEKIbJheZAYCEWSmEGSiUDlbiNm3NpQZ0nA0w794685WSZhsWAxZv2RBaYmE9bkRd170oulhLfjNPkibjMXQXDsCGv3g3+ALQS0PJNhGL4ps690ZM4VQcxAG5i/h/N0MnBtNRMloOkLaafAbuMjo1UKhPvx6PbZVsiodEoPpo3pCesQfTljVMD9tRNXESeitsUjtUSBp6DgGn1xH20wTBDVcMFrOrU3RJd2iOdokAm0UwfKsFZXjKSi0xCDVFAZtXwyuPqiFfkCLa9N14FfywCv3XPIp5uRvgog7hCtJg3Ew3MuHbkiFlC4FErqP4cpMDXRmDVILIqHLjIX6rBKBZQJ46zjwzuNgPcVl4ddsawZapkuReiMC8V1KGMbLkNuZhCjGFzWObEy52KHIdT/8/APeiAAHTpZ7xW8A3Sy6IDTwUT1RigSjCoaxcpw2qeCnpBDvuBPv7Xeg3YWFfmenNfPefTA6sx/r2tr8N9UQ1gpmOUVu6LzfimpbHrKMKshiFJg46IxHLIJ3hIVBQqDfvWcVu9gwuLh++WsL/HLurKCchzOmbOi7c+ERdQQlDmzME/JjisXCTUKQ43oAFWGhyPJzX9pyUFy9V7dXvvuaZ447uNEUkr098YmQ7y8JWQtxO7TYrE6sS+fx6tXtVfItARtF5VPzyRolXhGCXlc2ZFHhLf99+lMead6ridFe31T2Ts3hKenBhBCSLpc7ajQau5+bn1TH+A/7owAAAABJRU5ErkJggg==",
        "frecency": 2000,
        "lastVisitDate": 1457558902213,
        "parsedUrl": {
          "hash": "",
          "query": {},
          "protocol": "https:",
          "pathname": "/show_bug.cgi",
          "auth": "",
          "host": "bugzilla.mozilla.org",
          "port": "",
          "hostname": "bugzilla.mozilla.org",
          "password": "",
          "username": "",
          "href": "https://bugzilla.mozilla.org/show_bug.cgi"
        },
        "title": "1255136 – Deploy TestPilot 2016-03-09",
        "type": "history"
      },
      {
        "url": "https://www.npmjs.com/package/react-json-tree",
        "app_links": [],
        "authors": [],
        "cache_age": 86400,
        "content": null,
        "description": "React JSON Viewer Component, Extracted from redux-devtools",
        "embeds": [],
        "entities": [...],
        "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQBAMAAADt3eJSAAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAAhUExURcwAAP76+txLS////9tJSeFlZfri4tIYGNEQENxMTO2hobE3/qkAAAA2SURBVAjXY2DABC1KQOABZDgKAoEIkKFoDARCYIZZ6GIooyCYZIajoFhBoAjYiikMlR5Y7AYAO2YP82tQyKUAAAAASUVORK5CYII=",
        "favicon_colors": [...],
        "favicon_url": "https://www.npmjs.com/static/images/touch-icons/favicon.ico",
        "frecency": 100,
        "images": [...],
        "keywords": [...],
        "language": "English",
        "lastVisitDate": 1457557402100,
        "lead": null,
        "media": {},
        "offset": null,
        "original_url": "https://www.npmjs.com/package/react-json-tree",
        "parsedUrl": {
          "hash": "",
          "query": {},
          "protocol": "https:",
          "pathname": "/package/react-json-tree",
          "auth": "",
          "host": "www.npmjs.com",
          "port": "",
          "hostname": "www.npmjs.com",
          "password": "",
          "username": "",
          "href": "https://www.npmjs.com/package/react-json-tree"
        },
        "provider_display": "www.npmjs.com",
        "provider_name": "npm",
        "provider_url": "https://www.npmjs.com",
        "published": null,
        "related": [],
        "safe": true,
        "title": "react-json-tree",
        "type": "history"
      },
      {
        "url": "https://search.yahoo.com/yhs/search?p=react+json+tree&ei=UTF-8&hspart=mozilla&hsimp=yhs-002",
        "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAADA0lEQVQ4jV2RS0xcZRiG3/l7zn9Ku4F127hyqUFnzthTGFJoGaZOmCmkHVqKMdq68EKRAGFahVKSUsCApoPdcDOOYoSZJk2KXVgSKaVNmhiN2jQYoxvZGRftmnlcIFPjl7z57pc3n8LqJKIuwupaC6vnnK8PDlRrqHIHtcpW1epqVbWGKl/R0P6Ius+G1fVgu6cTRZUlpo9o0BRx+ymHQ5PEzCgxM0qtuUKtuVL2D4cmaVCOBk0R0zhRZVGgS8T1GcdDN/nzt78BSDsFUs4Sze7XpNxFUs4Sx20RSvD7L3+R1k0aNUegSyjQMEkt0eZ8x4ev32NHTtu7nLKrnLKrtHt3y/GBN++RMasktUSgYRToMsnQIm1mjdfchzz64Q8Apt59RId3nw7vPhNv/ATA52Pf0+E8pM2s8WroKwJdRoEGnqRMkZPOCu3uOmfssyvOeOt07F5/5tt12t11TjorpEyBQANPdFAXN5vMPK3ON2Tst7R5d5gdfgxAfvxXPnnrZ0olmB1+TJu3QsbeocVZpsnMc1AXNxVVdqPJzJN2i7TaW5zwlsnsuc1/pVSCzJ7bnPCWabW3SLkFmsw8UWU3FFV2I26mSboLpG2BFq9Ia0WxzBtg7MyPtFbcoMUrkLYFku6XxM309gBf/ZsNJkfCmSdp8zR7edIVC5x+vri9fatE5rlF0hULNHt5kjZPwpmjwVzDV/+mIup/WmcmaHSuE3enSdgZjnkzJPfOAbC1VSK5d45j3iwJO0PcnabRuU6dmSCi/qfy1UdNaIR68zFHnBxH7TXiNkfcmypTiHtTxG2OozbHESdHvZmkJjSCrz4UUQ+BhqgxI8R2jVLnjFHvjtNckS9TaN79BfXuGHXOGLFdo9SYqwQaIqIeFFY3UV0g0ACBGSQwgxxyB0nsmyx/ILFvkkPOYDkfaJCoLhBWNwqrC1+9+OrDVx/RUC++6SX6P/imdzv3b52vXsLqQmF1so3za2GdP+frnQPVer/yBb1dtaN37BfVuf8lvXf2ZXU+2On7B+kzEw3V663qAAAAAElFTkSuQmCC",
        "frecency": 100,
        "lastVisitDate": 1457557399270,
        "parsedUrl": {
          "hash": "",
          "query": {},
          "protocol": "https:",
          "pathname": "/yhs/search",
          "auth": "",
          "host": "search.yahoo.com",
          "port": "",
          "hostname": "search.yahoo.com",
          "password": "",
          "username": "",
          "href": "https://search.yahoo.com/yhs/search"
        },
        "title": "react json tree - - Yahoo Search Results",
        "type": "history"
      }
    ],
    "error": false
  },
```

**TL;DR:** @pdehaan screwed up. Don't let him touch your code.

**Comment by pdehaan** [comment]
Turns out, this is partially due to `urlParse()` not parsing the query string as an object in the [**parse-url-middleware.js:18**](https://github.com/mozilla/activity-streams/blob/0c1ce94335080a5dbcc8e7fbce7bde01cd81bbd4/content-src/lib/parse-url-middleware.js#L18);
```js
const parsedUrl = urlParse(site.url, false);
```

But in the [**/content-src/lib/utils.js:29** `sanitizeUrl()`](https://github.com/mozilla/activity-streams/blob/0c1ce94335080a5dbcc8e7fbce7bde01cd81bbd4/content-src/lib/utils.js#L29) method, we expect to filter the query params and loop over them:
```js
const parsedUrl = (typeof site === "string") ? urlParse(site, true) : site.parsedUrl;
```

That still doesn't explain why it isn't populating the Yahoo! query though.

```json
        {
          "url": "https://search.yahoo.com/yhs/search?p=react+json+tree&ei=UTF-8&hspart=mozilla&hsimp=yhs-002",
          "title": "react json tree - - Yahoo Search Results",
          "lastVisitDate": 1457572674302,
          "frecency": 100,
          "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAADA0lEQVQ4jV2RS0xcZRiG3/l7zn9Ku4F127hyqUFnzthTGFJoGaZOmCmkHVqKMdq68EKRAGFahVKSUsCApoPdcDOOYoSZJk2KXVgSKaVNmhiN2jQYoxvZGRftmnlcIFPjl7z57pc3n8LqJKIuwupaC6vnnK8PDlRrqHIHtcpW1epqVbWGKl/R0P6Ius+G1fVgu6cTRZUlpo9o0BRx+ymHQ5PEzCgxM0qtuUKtuVL2D4cmaVCOBk0R0zhRZVGgS8T1GcdDN/nzt78BSDsFUs4Sze7XpNxFUs4Sx20RSvD7L3+R1k0aNUegSyjQMEkt0eZ8x4ev32NHTtu7nLKrnLKrtHt3y/GBN++RMasktUSgYRToMsnQIm1mjdfchzz64Q8Apt59RId3nw7vPhNv/ATA52Pf0+E8pM2s8WroKwJdRoEGnqRMkZPOCu3uOmfssyvOeOt07F5/5tt12t11TjorpEyBQANPdFAXN5vMPK3ON2Tst7R5d5gdfgxAfvxXPnnrZ0olmB1+TJu3QsbeocVZpsnMc1AXNxVVdqPJzJN2i7TaW5zwlsnsuc1/pVSCzJ7bnPCWabW3SLkFmsw8UWU3FFV2I26mSboLpG2BFq9Ia0WxzBtg7MyPtFbcoMUrkLYFku6XxM309gBf/ZsNJkfCmSdp8zR7edIVC5x+vri9fatE5rlF0hULNHt5kjZPwpmjwVzDV/+mIup/WmcmaHSuE3enSdgZjnkzJPfOAbC1VSK5d45j3iwJO0PcnabRuU6dmSCi/qfy1UdNaIR68zFHnBxH7TXiNkfcmypTiHtTxG2OozbHESdHvZmkJjSCrz4UUQ+BhqgxI8R2jVLnjFHvjtNckS9TaN79BfXuGHXOGLFdo9SYqwQaIqIeFFY3UV0g0ACBGSQwgxxyB0nsmyx/ILFvkkPOYDkfaJCoLhBWNwqrC1+9+OrDVx/RUC++6SX6P/imdzv3b52vXsLqQmF1so3za2GdP+frnQPVer/yBb1dtaN37BfVuf8lvXf2ZXU+2On7B+kzEw3V663qAAAAAElFTkSuQmCC",
          "type": "history",
          "parsedUrl": {
            "hash": "",
            "query": {
              "p": "react+json+tree"
            },
            "protocol": "https:",
            "pathname": "/yhs/search",
            "auth": "",
            "host": "search.yahoo.com",
            "port": "",
            "hostname": "search.yahoo.com",
            "password": "",
            "username": "",
            "href": "https://search.yahoo.com/yhs/search?p=react%2Bjson%2Btree"
          }
        }
```

**Comment by k88hudson** [comment]
This should be closed by #304, just keeping the ticket around for verification purposes.

**Comment by sarracini** [comment]
#304 landed, closing this issue
